### PR TITLE
modify the API to use smart pointers to pave the way for the async

### PR DESCRIPTION
**Pull request by Presburger**
The async interface needs to hold the config object even after the API call has ended.

### CODE REVIEW

**Comment by Presburger** [comment]
issue: #829

File: include/knowhere/index/index_node.h [hunk 1]
```diff
@@ -62,9 +62,9 @@ class IndexNode : public Object {
      * to be revisited.
      */
     virtual Status
-    Build(const DataSetPtr dataset, const Config& cfg) {
+    Build(const DataSetPtr dataset, std::shared_ptr<Config> cfg) {
```
Comment (alexanderguzhva) [inline include/knowhere/index/index_node.h:65]: Please add comments to all methods on why it is needed to pass `Config` this way. It is clear from the description of this PR, but it not very clear without it. 
Please also explain why you pick `std::shared_ptr<>` or `std::unique_ptr<>` depending on a use case.

**Comment by alexanderguzhva** [comment]
/lgtm 
overall

**Pull request merged**